File: packages/app-shortcuts/android/src/main/java/io/capawesome/capacitorjs/plugins/appshortcuts/AppShortcutsPlugin.java
package io.capawesome.capacitorjs.plugins.appshortcuts;

import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.appshortcuts.classes.events.ClickEvent;
import io.capawesome.capacitorjs.plugins.appshortcuts.classes.options.SetOptions;
import io.capawesome.capacitorjs.plugins.appshortcuts.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.appshortcuts.interfaces.NonEmptyCallback;
import io.capawesome.capacitorjs.plugins.appshortcuts.interfaces.Result;
import java.util.Objects;
import org.json.JSONArray;
import org.json.JSONObject;

@CapacitorPlugin(name = "AppShortcuts")
public class AppShortcutsPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String ERROR_SHORTCUTS_MISSING = "shortcuts must be provided.";
    public static final String ERROR_TITLE_MISSING = "title must be provided.";
    public static final String ERROR_ID_MISSING = "id must be provided.";
    public static final String EVENT_CLICK = "click";
    public static final String INTENT_EXTRA_ITEM_NAME = "shortcutId"; // DO NOT CHANGE THIS VALUE!
    public static final String TAG = "AppShortcutsPlugin";

    private AppShortcuts implementation;

    @Override
    public void load() {
        this.implementation = new AppShortcuts(getContext(), getAppShortcutsConfig());
    }

    @PluginMethod
    public void get(PluginCall call) {
        try {
            NonEmptyCallback<Result> callback = new NonEmptyCallback<>() {
                @Override
                public void success(@NonNull Result result) {
                    resolveCall(call, result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.get(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void set(PluginCall call) {
        try {
            SetOptions options = new SetOptions(call, getContext(), getBridge());
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call, null);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.set(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void clear(PluginCall call) {
        try {
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call, null);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.clear(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void notifyClickListener(@NonNull ClickEvent event) {
        notifyListeners(EVENT_CLICK, event.toJSObject(), true);
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable JSObject result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    @Override
    protected void handleOnNewIntent(Intent intent) {
        super.handleOnNewIntent(intent);
        if (Objects.equals(intent.getAction(), Intent.ACTION_VIEW)) {
            String shortcutId = intent.getStringExtra(AppShortcutsPlugin.INTENT_EXTRA_ITEM_NAME);
            if (shortcutId != null) {
                ClickEvent event = new ClickEvent(shortcutId);
                this.notifyClickListener(event);
            }
        }
    }

    private AppShortcutsConfig getAppShortcutsConfig() {
        AppShortcutsConfig config = new AppShortcutsConfig();
        JSONObject configJSON = getConfig().getConfigJSON();
        try {
            JSONArray shortcutsJSON = configJSON.getJSONArray("shortcuts");
            JSArray shortcuts = new JSArray();
            for (int i = 0; i < shortcutsJSON.length(); i++) {
                shortcuts.put(shortcutsJSON.get(i));
            }
            config.setShortcuts(AppShortcutsHelper.createShortcutInfoCompatList(shortcuts, getContext(), getBridge()));
        } catch (Exception e) {
            return config;
        }
        return config;
    }
}

File: packages/app-shortcuts/ios/Plugin/AppShortcutsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(AppShortcutsPlugin)
public class AppShortcutsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppShortcutsPlugin"
    public let jsName = "AppShortcuts"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise)
    ]
    public static let notificationName = "handleAppShortcutNotification"
    public static let userInfoShortcutItemKey = "shortcutItem"

    public let eventClick = "click"
    public let tag = "AppShortcuts"

    private var implementation: AppShortcuts?

    override public init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(self.handleAppShortcutNotification), name: NSNotification.Name(AppShortcutsPlugin.notificationName), object: nil)
    }

    override public func load() {
        super.load()
        self.implementation = AppShortcuts(getAppShortcutConfig())
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc func get(_ call: CAPPluginCall) {
        implementation?.get(completion: { result in
            if let result = result.toJSObject() as? JSObject {
                self.resolveCall(call, result)
            }
        })
    }

    @objc func set(_ call: CAPPluginCall) {
        do {
            let options = try SetOptions(call: call)
            implementation?.set(shortcuts: options.getShortcuts, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, nil)
                }
            })
        } catch let error {
            rejectCall(call, error)
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        implementation?.clear(completion: { error in
            if let error = error {
                self.rejectCall(call, error)
            } else {
                self.resolveCall(call, nil)
            }
        })
    }

    @objc private func handleAppShortcutNotification(notification: Notification) {
        guard let userInfo = notification.userInfo, let shortcutItem = userInfo[AppShortcutsPlugin.userInfoShortcutItemKey] as? UIApplicationShortcutItem else {
            return
        }
        let event = ClickEvent(shortcutItem)
        notifyDeviceScannedListeners(event: event)
    }

    func notifyDeviceScannedListeners(event: ClickEvent) {
        if let event = event.toJSObject() as? JSObject {
            notifyListeners(eventClick, data: event, retainUntilConsumed: true)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", self.tag, "] ", error)
        call.reject(error.localizedDescription)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: JSObject?) {
        if let result {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }

    private func getAppShortcutConfig() -> AppShortcutsConfig {
        var config = AppShortcutsConfig()
        if let shortcuts = getConfig().getArray("shortcuts") as? [JSObject] {
            do {
                config.shortcuts = try AppShortcutsPluginHelper.getShortcutItemsFromJSArray(shortcuts: shortcuts)
            } catch {
                return config
            }
        }
        return config
    }
}
